sftp: add options for subsystem and server_command - fixes #1801

aus opened the pull request:
#### What is the purpose of this change?

The traditional OpenSSH `sftp` command supports specifying the SSH subsystem or the SFTP server via the `-s` option. This PR adds two options to the sftp backend: `subsystem` and `server_command`. 

The `subsystem` option is useful if the target server has multiple SFTP subsystems defined for users.

The `server_command` option can be useful if `sudo` is required to elevate the user sftp session, without the need to SSH as root user.

The new `newSftpClient()` function is essentially a just reimplementation of [sftp.NewClient](https://godoc.org/github.com/pkg/sftp#NewClient) to support custom subsystem or execution of a custom SFTP server on the remote host.

#### Was the change discussed in an issue or in the forum before?

It was briefly discussed in 2017 in issue #1801.

#### Checklist

- [x] I have read the [contribution guidelines](https://github.com/rclone/rclone/blob/master/CONTRIBUTING.md#submitting-a-pull-request).
- [x] I have added tests for all changes in this PR if appropriate.
- [x] I have added documentation for the changes if appropriate.
- [x] All commit messages are in [house style](https://github.com/rclone/rclone/blob/master/CONTRIBUTING.md#commit-messages).
- [x] I'm done, this Pull Request is ready for review :-)

ncw commented on backend/sftp/sftp.go:
```diff
@@ -164,6 +164,16 @@ Home directory can be found in a shared folder called "home"
+		}, {
+			Name:     "sftp_server",
+			Default:  "",
+			Help:     "Specifies the path or command to run a sftp server on the remote host. The subsystem option is ignored when sftp_server is defined.",
```
Can you break this line into two parts - the first line makes the option help text for `--sftp-server` and its too long!

`Specifies the path or command to run a sftp server on the remote host.

The subsystem option is ignored when sftp_server is defined.`

ncw commented on backend/sftp/sftp.go:
```diff
@@ -164,6 +164,16 @@ Home directory can be found in a shared folder called "home"
+			Help:     "Specifies the SSH2 subsystem on the remote host.",
+			Advanced: true,
+		}, {
+			Name:     "sftp_server",
```
Can we call this something other than `sftp_server` otherwise the option will be called `--sftp-sftp-server`! Maybe `server` or `server_command`?

ncw commented on backend/sftp/sftp.go:
```diff
@@ -187,6 +197,8 @@ type Options struct {
 	Sha1sumCommand    string `config:"sha1sum_command"`
 	SkipLinks         bool   `config:"skip_links"`
+	Subsystem         string `config:"subsystem"`
+	SftpServer        string `config:"sftp_server"`
```
Don't forget to rename this too!

ncw commented on backend/sftp/sftp.go:
```diff
@@ -299,6 +311,35 @@ func (f *Fs) sftpConnection() (c *conn, err error) {
 	return c, nil
 }
 
+// Creates a new SFTP client on conn, using specificed subsystem
```
s/specificed/the specified/

aus commented:
Great feedback. I've implemented your suggestions and squashed the commits. (Ignore the first two erroneous force-pushes 😅)

Also, just wanted to say that I've been using rclone for years, and it's awesome! Thanks for all your work.

ncw reviewed (approved):
This looks great now thanks. I'll merge it now.

ncw commented:
> Great feedback. I've implemented your suggestions and squashed the commits. (Ignore the first ~two~ three erroneous force-pushes

😃

> Also, just wanted to say that I've been using rclone for years, and it's awesome! Thanks for all your work.

Thank you, and thank you for your contribution 😃

smac89 commented:
@aus 
This is not working the way it was originally stated in #1801.

I thought the original intent by raising that issue was to allow `stfp-server` to be run with sudo. However, that does not seem to be the case. I've tried all kinds of combinations and the only option that seems to work is to just specify `sftp` as the subsystem, not even a path like `/usr/lib/openssh/sftp-server`, works.

I was under the impression that with this PR, we can now supply something like `sudo /usr/lib/openssh/sftp-server` for the subsystem option (which btw is allowed when running sftp from the commandline)

smac89 commented:
Ignore the above. I just read the PR again, and it looks like I need to be using `server_command` instead of `subsystem`
